dbms/src/Interpreters/Join.cpp

```diff
@@ -2337,13 +2341,78 @@ void Join::checkTypes(const Block & block) const
     checkTypesOfKeys(block, sample_block_with_keys);
 }
 
+Join::MaterializedNullRows::MaterializedNullRows()
+    : null_list(nullptr)
+    , null_list_pos(0)
+    , null_list_it(nullptr)
+    , materialized_rows(0)
+{}
+
+void Join::MaterializedNullRows::setRowList(const std::vector<Join::RowsNotInsertToMap> * null_list_)
+{
+    null_list = null_list_;
+}
+
+bool Join::MaterializedNullRows::fillColumns(MutableColumns & added_columns, size_t left_columns, size_t right_columns, size_t & pos, size_t max_pace)
+{
+    RUNTIME_ASSERT(added_columns.size() >= left_columns + right_columns);
+
+    if (unlikely(max_pace == 0))
+        return false;
+
+    if (unlikely(materialized_columns.empty()))
+    {
+        RUNTIME_ASSERT(right_columns > 0);
+        materialized_columns.resize(right_columns);
+        for (size_t j = 0; j < right_columns; ++j)
+            materialized_columns[j] = added_columns[j + left_columns]->cloneEmpty();
+    }
+
+    bool end = false;
+    while (materialized_rows < pos + max_pace)
+    {
+        while (null_list_it == nullptr)
+        {
+            if (null_list_pos >= null_list->size())
+            {
+                end = true;
+                break;
+            }
+            ++null_list_pos;
+            null_list_it = (*null_list)[null_list_pos - 1].head.next;
+        }
+        if (end)
+            break;
+
+        for (size_t j = 0; j < right_columns; ++j)
+            materialized_columns[j]->insertFrom(*null_list_it->block->getByPosition(j).column.get(), null_list_it->row_num);
+
+        ++materialized_rows;
+        null_list_it = null_list_it->next;
+    }
+
+    if (materialized_rows <= pos)
+    {
+        /// No more rows should be filled.
+        return true;
+    }
+    size_t pace = std::min(materialized_rows - pos, max_pace);
+
+    for (size_t j = 0; j < right_columns; ++j)
+        added_columns[j + left_columns]->insertRangeFrom(*materialized_columns[j].get(), pos, pace);
```

Comment:
It seems the cached materialized_columns would be read only once, would it really help improve performance?

Reply:
`materialized_columns` may be used by other rows in left table, which will benefit from cached `materialized_columns`.
By the way, I change this implementation because memory usage is `materialized_columns` times the number of threads, which can be very large.

Now I change to materialize the null rows during build phase. Therefore the extra memory overhead is `materialized_columns`.